### PR TITLE
Refactor Client Code, part 1 of several

### PR DESCRIPTION
**Pull request by brandonforster**
This PR fixes https://github.com/edgexfoundry/edgex-go/issues/2237.

It is the first part of several working on #196.

While addressing the root cause of that issue, it was realized that the bug was not contained solely to the `device` endpoint, nor was it in command. The problem lies in how every endpoint that is exposed for cross service communication initializes its URL field.

This PR addresses the issue on metadata's `device` client, fixing 2237. It also lays out the strategy that will eventually be applied to every endpoint and the structures and objects that will be used to accomplish that. They were not added to this PR for the sake of review. Any feedback on this PR will be used to inform the future PRs.

To test, actuate the endpoints in the metadata device client. The blackbox tests for command should work for that.

### CODE REVIEW

**Comment by michaelestrin** [comment]
recheck

File: clients/interfaces/builder.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package interfaces
+
+// RestClientBuilder is the interface for types that need to define some way to retrieve state information about services.
+// This information can be anything that must be determined at runtime, whether it is unknown or simply not yet known.
+type RestClientBuilder interface {
```
Comment (michaelestrin) [inline clients/interfaces/builder.go]: Name is ambiguous to me.  "builder" is a relatively generic term.  Could be mistaken for the GoF Builder pattern (which it most definitely isn't a specific implementation of).

Could this be simply a URL interface?

File: clients/metadata/device.go [hunk 2]
```diff
@@ -17,11 +17,13 @@ package metadata
 import (
 	"context"
 	"encoding/json"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/rest"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/types"
```
Comment (michaelestrin) [inline clients/metadata/device.go]: Sort imports.

File: clients/metadata/device.go [hunk 3]
```diff
@@ -74,35 +76,22 @@ type DeviceClient interface {
 }
 
 type deviceRestClient struct {
-	url      string
-	endpoint clients.Endpointer
+	client interfaces.RestClientBuilder
```
Comment (michaelestrin) [inline clients/metadata/device.go]: Match field name to contract name; i.e. RestClientBuilder type -> restClientBuilder (or builder) field name.

Commented once; may apply elsewhere.

File: clients/rest/consul.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+import (
+	"errors"
+	"time"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/interfaces"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/types"
+)
+
+type consulClient struct {
```
Comment (michaelestrin) [inline clients/rest/consul.go]: Add docblock.

File: clients/rest/consul.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	if c.initialized {
+		return c.url, nil
+	}
+
+	timer := time.After(time.Duration(c.timeout) * time.Second)
+	ticker := time.Tick(500 * time.Millisecond)
+	for {
+		select {
+		case <-timer:
+			return "", errors.New("client not yet initialized")
```
Comment (michaelestrin) [inline clients/rest/consul.go]: Export error text into public constant.  More accurate description might be "unable to initialize client in %c.timeout seconds"

File: clients/rest/consul.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
+/*******************************************************************************
```
Comment (michaelestrin) [inline clients/rest/consul.go]: Create corresponding consul_test.go with unit tests.

File: clients/rest/factory.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
+/*******************************************************************************
```
Comment (michaelestrin) [inline clients/rest/factory.go]: Create corresponding factory_test.go with unit tests.

File: clients/rest/local.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
+/*******************************************************************************
```
Comment (michaelestrin) [inline clients/rest/local.go]: Create corresponding local_test.go with unit tests.

File: clients/rest/local.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+ * the License.
+ *******************************************************************************/
+
+package rest
+
+import (
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/types"
+)
+
+type localClient struct {
```
Comment (michaelestrin) [inline clients/rest/local.go]: Add docblock.

File: clients/metadata/device.go [hunk 10]
```diff
@@ -20,8 +20,11 @@ import (
 	"net/url"
 	"strconv"
 
-	"github.com/edgexfoundry/go-mod-core-contracts/clients"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/rest"
 	"github.com/edgexfoundry/go-mod-core-contracts/clients/types"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/clients"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/interfaces"
```
Comment (michaelestrin) [inline clients/metadata/device.go]: Resort imports.  All are go-mod-core-contracts-based.  Empty line should be removed.

File: clients/rest/factory_test.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/types"
+)
+
+func TestClientFactoryLocal(t *testing.T) {
+	actualClient := ClientFactory(types.EndpointParams{UseRegistry: false}, nil)
```
Comment (michaelestrin) [inline clients/rest/factory_test.go]: In production usage, I believe ClientFactory's m parameter will always be non-nil.  Test should reflect that.

File: clients/rest/local_test.go [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+	expectedURL := "http://brandonforster.com"
+	actualClient := newLocalClient(types.EndpointParams{Url: expectedURL})
+
+	if actualClient.url != expectedURL {
+		t.Fatalf("expected URL %s, found URL %s", expectedURL, actualClient.url)
+	}
+}
+
+func TestLocalClient_URLPrefix(t *testing.T) {
+	expectedURL := "http://brandonforster.com"
```
Comment (michaelestrin) [inline clients/rest/local_test.go]: Use `http://domain.com` instead.

Commented once; occurs elsewhere.

File: clients/rest/local_test.go [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/types"
+)
+
+func TestNewLocalClient(t *testing.T) {
+	expectedURL := "http://brandonforster.com"
+	actualClient := newLocalClient(types.EndpointParams{Url: expectedURL})
+
+	if actualClient.url != expectedURL {
+		t.Fatalf("expected URL %s, found URL %s", expectedURL, actualClient.url)
+	}
```
Comment (michaelestrin) [inline clients/rest/local_test.go]: Tests shouldn't reference internal implementation.  `url` is private.  

I think the purpose of this test is to ensure `newLocalClient` returns a non-nil `*localClient` value.  Your assertion should be updated to reflect that.

File: clients/rest/registry_test.go [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+package rest
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/types"
+)
+
```
Comment (michaelestrin) [inline clients/rest/registry_test.go]: Add a test that ensures `newRegistryClient` returns a non-nil `*registryClient` value.

**Comment by brandonforster** [comment]
> The `rest` package at `clients/rest` isn't named well. It encapsulates a factory and different strategies for obtaining a client URL. `rest` doesn't describe this responsibility well.

Thanks. I've done my best to redo it according to https://blog.golang.org/package-names, let me know if you think there's still more to do there.

File: clients/urlclient/local_test.go [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+	if actualClient == nil {
+		t.Fatal("nil returned from newLocalClient")
+	}
+
+	expectedType := reflect.TypeOf(&localClient{})
+	clientType := reflect.TypeOf(actualClient)
+
+	if clientType != expectedType {
+		t.Fatalf("expected type %T, found %T", expectedType, actualClient)
+	}
```
Comment (michaelestrin) [inline clients/urlclient/local_test.go]: Reflection isn't necessary since `newRegistryClient` returns *registryClient test will never fail.

File: clients/urlclient/registry_test.go [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+	if actualClient == nil {
+		t.Fatal("nil returned from newRegistryClient")
+	}
+
+	expectedType := reflect.TypeOf(&registryClient{})
+	clientType := reflect.TypeOf(actualClient)
+
+	if clientType != expectedType {
+		t.Fatalf("expected type %T, found %T", expectedType, actualClient)
+	}
```
Comment (michaelestrin) [inline clients/urlclient/registry_test.go]: Reflection isn't necessary since `newRegistryClient` returns *registryClient test will never fail.

File: clients/urlclient/local_test.go [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+func TestLocalClient_URLPrefix(t *testing.T) {
+	expectedURL := "http://domain.com"
+	client := newLocalClient(types.EndpointParams{Url: expectedURL})
+
+	actualURL, err := client.Prefix()
```
Comment (michaelestrin) [inline clients/urlclient/local_test.go]: With rename of `localClient.URLPrefix()` to `localClient.Prefix`, `client` here should be renamed here to `urlClient` to match the changes made elsewhere (and more importantly, to provide context for what `Prefix()` actually returns).

**Comment by michaelestrin** [comment]
Merged with https://github.com/edgexfoundry/go-mod-core-contracts/pull/197
